### ISSUE TITLE
Calculate inputs once during plan construction?

### ISSUE THREAD

**Issue by gmbecker** [ropensci/drake#504]
@wlandau it looked from your talk that the plan is a data.frame that has the target-code linking but it doesn't have the the inputs for the expression. Those aren't going to change for as long as the expression doesn't change, so why not calculate it up front and put it in the plan df right away. You can use a list column for this to get around the fact that an expression can have 0-many inputs.  Apologies if I missed something under the hood that avoids this repetition already under the hood.

**Comment by wlandau** [comment]
Glad we got a chance to talk about this in person yesterday. I got pretty far along in development before I realized code analysis was a bottleneck, and `make()` indeed repeats the dependency detection step in most cases. In the current `drake`, there are a couple existing ways to speed things along:

1. Pre-calculate a `drake_config()` object and pass it to the `config` argument of `make()`, or
2. Pre-calculate the dependency graph with `build_drake_graph()` and pass it to the `graph` argument of `drake_config()` or `make()`. This graph is an `igraph` of targets/jobs with specific dependency information stored in vertex attributes.

These approaches avoid the redundant computation of not only the code analysis, but also the construction of the dependency/job `igraph` afterwards, both of which can be bottlenecks for jobs with lots of targets. However, if the user changes an imported function enough to change the dependency relationships, they will need a new `drake_config()` or `igraph` object all over again. With your approach, users would still be free to change custom functions, and we could still save time during the initial stages of `make()`.

So far, I see a couple ways to do this, and there may be more.

#### `finalize_plan()`

We could implement your proposed `finalize_plan()` function to add each target's dependency information as a list column in the plan. Users would construct the usual plan with `drake_plan()` or some custom method, call `finalize_plan()`, and then call `make()`. That way, users would be free to change their custom (non-command) functions and still save time on the preprocessing. While this does seem to strike a great balance between time savings and flexibility for long plans, I am not sure about adding another required step in the workflow for everyone.

#### A new `deps` argument of `make()`

We could optionally pass the list of target dependencies into a new `deps` argument of `make()`/`drake_config()`. In my current thinking, `deps` would behave just like the `graph` and `config` arguments. The default behavior of `make()` would not change, but power users could supply a `deps` list with all the pre-calculated dependencies of the targets. We could have a new `deps_plan()` function to create this dependency list and group it alongside `deps_code()` and `deps_targets()` in the documentation. 

That's the straightforward part of this approach. We would also need to refactor [`build_drake_graph()`](https://github.com/ropensci/drake/blob/master/R/build_drake_graph.R#L28) and [`bdg_analyze_code()`](https://github.com/ropensci/drake/blob/master/R/build_drake_graph.R#L78), the latter of which should split into a function for imports and a function for targets, and pass in the optional `deps` argument. And then there's the question of whether we also do this for [triggers](https://ropenscilabs.github.io/drake-manual/triggers.html) because the optional `change` and `depends` triggers also have code.

#### My overall thoughts so far

Exposing the code analysis stage at the `drake_plan()` level has advantages for both performance and reproducibility. I am not sure yet how much value is added relative to the current `graph` and `config` arguments, especially when weighed against the ways the API and internals would become more complicated. `drake` is already much larger and much more complicated than I ever intended it to be, but at the same time, there is a need to follow through on `drake`'s promises of fast and reproducible computation.

**Comment by gmbecker** [comment]
@wlandau I agree it was great to meet in person and chat about this. Sorry I had to run to that dinner and cut the conversation a bit short.

Another thing to consider is whether we can make a drake plan know whether it has been finalized or not. If so, make could easily call finalize on it's input if the plan isn't already finalized, thus giving us the best of both worlds.

Now because you're entirely in S3 land, encoding this knowledge in a completely safe way may be a bit tricky, but is ultimately doable, I think. The way I would probably do this is via something like the following (I know I know, I'm lame with my `=` assignments and lack of pipes) 

```r
finalize_plan = function(plandf) {
    plandf$deps = .make_deps_column(plan)
    attr(plandf, "finalized") = NULL
    hash = fastdigest(plandf[, 1:3]) # the subset gets rid of the attributes
    attr(plandf, "finalized") = hash
    plandf
}

drake_is_finalized = function(plandf) {
    attrval = attr(plandf, "finalized")
    ret = FALSE
    if( !is.null(attrval) && attrval == fastdigest(plandf[,1:3]))
        ret = TRUE
    ret
}

## I don't know the arguments to make so adapt as necessary
make = function(plandf, ...) {
    if(!drake_is_finalized(plandf))
        plandf = finalize_plan(plandf)
    ## the normal make code...
}
```

**Comment by wlandau** [comment]
I like what this approach is getting at, and I like the use of `fastdigest`.

The code analysis of `plandf` happens in [`bdg_analyze_commands()`](https://github.com/ropensci/drake/blob/1b6b3d0688652031fc00ae5fc689465e7d1965c0/R/build_drake_graph.R#L102), one of several steps in [`build_drake_graph()`](https://github.com/ropensci/drake/blob/master/R/build_drake_graph.R#L29).

https://github.com/ropensci/drake/blob/1b6b3d0688652031fc00ae5fc689465e7d1965c0/R/build_drake_graph.R#L55-L62

Upon further reflection, I think it would be nice to have the same solution apply to each of these preprocessing steps. I think memoization really fits, particularly if we use a `storr` namespace in `drake`'s existing cache. Unfortunately, that would exclude `fastdigest`, but hopefully https://github.com/richfitz/storr/issues/82 will open up new performance gains for `drake` more broadly.

**Comment by wlandau** [comment]
Edit: I believe @krlmlr suggested an approach like this at some point. I was not ready to follow through with a solution back then, but with the recent refactoring I did on `build_drake_graph()`, I think the implementation will be easier this time around.

**Comment by wlandau** [comment]
It just occurred to me that in the current implementation of `build_drake_graph()`, traditional memoization may not exactly fit the `bdg_` functions because they all take an all-inclusive `args` list. In `bdg_analyze_commands()` and `bdg_analyze_triggers()`, we might try something more like https://github.com/ropensci/drake/issues/504#issuecomment-414374511 to react to just the changes in the relevant columns of the plan. For other functions, this custom memoization-like step would be similar. The existing `storr` cache should make this straightforward.

**Comment by wlandau** [comment]
Fixed via #507.
